More options for hawk/dove neighborhoods

rlskoeser opened the pull request:
@LaraBuchak this implements the new variations in neighborhood sizes you've described.  I've added a third size (24), adjusted how neighborhood sizes are configured, and renamed these based on the different neighborhoods you described. I've named them `play_neighborhood` (neighbors an agent plays against), `choose_neighborhood` (neighbors an agent looks at when choosing what to play for the next round) and `adjust_neighborhood` (neighbors an agent looks at when adjusting risk attitude).

Once I went to implement it I understood your notes about the choose neighborhood - because our risk strategies are discrete values this can't go higher than 8 with current logic. I've set it up so the `choose_neighborhood` size matches the `play_neighborhood` size but won't go higher than 8 (i.e. if you use a play neighborhood of 24 you get a choose neighborhood of 8).

I put this version of the code up on our server so you can try it out. You should see that the neighborhood options have changed for both single and variable risk versions of the hawk/dove games.

https://simrisk.pulcloud.io/hawkdove-single 
https://simrisk.pulcloud.io/hawkdove-variable

quadrismegistus commented on simulatingrisk/hawkdove/model.py:
```diff
@@ -66,18 +66,38 @@ def initial_choice(self, hawk_odds=None):
 
+    def get_neighbors(self, size):
+        """get all neighbors for a supported neighborhood size"""
+        check_neighborhood_size(size)
```
maybe a comment here that this could raise a `ValueError` exception

quadrismegistus commented on simulatingrisk/hawkdove/server.py:
```diff
@@ -26,7 +31,7 @@ def agent_portrayal(agent):
 
     # color based on risk level
-    if agent.model.include_diagonals:
+    if agent.model.play_neighborhood >= 8:
```
Nice

quadrismegistus commented on simulatingrisk/hawkdove/model.py:
```diff
@@ -153,19 +173,25 @@ class HawkDoveModel(mesa.Model):
+        # since we define risk attitudes from 0-8 or 0-4,
+        # the neighborhood for choosing what to play cannot be bigger than 8;
+        # use the smaller of the play neighborhood or 8
+        self.choose_neighborhood = min(8, play_neighborhood)
```
this attribute name sounds like a function, but it's clear enough here

rlskoeser replied:
oh wow, it totally does - I didn't think of the fact that it sounds like a verb. I think it's probably ok in context, but I'll keep in mind if Lara or I can come up with a better name. (She had much longer descriptions in her write-up that I was trying to condense)

LaraBuchak replied:
On this, I think the neighborhood that determines what the agent should play should be 8 (or 4 in the "don't include diagonals" case), even if the play neighborhood is smaller. The intuitive idea behind the "choose neighborhood" is that the agent looks at it to determine the probability of a random agent playing HAWK, and uses that in her future interactions. It doesn't need to overlap with the agents she's actually playing, and there's an interesting case in which this probability is determined by a *larger* sample than the one she interacts with, e.g., she uses 8 neighbors to determine the probability but only interacts with 4. (You can imagine, for example, making predictions about what someone playing a game will do, even if you've never played with that particular person, by thinking about what's happened when you've played that game in the past with different people.) Happy to explain more if this doesn't make sense.

This is also where the assumption that Kevin called "myopia" entered in. A different way for an agent to determine the probability that some randomly-selected agent plays HAWK might be to look at what her 8 neighbors have done on every round so far, or on the last 10 rounds, or whatever. (Another potential thing to try, though it wasn't a priority--I'll add it to our list.)

Maybe it would be clearer if we called this "evidence neighborhood," since it's the neighborhood that serves as evidence for the agent to set her probabilities.

rlskoeser replied:
@LaraBuchak thanks for responding. I wondered about calling it "observed neighborhood" (relates to seen / myopia) but wondered if that was ambiguous with the neighbors used for deciding how to adjust risk attitudes, which is a different kind of observation.

I have a sense of the potential interesting cases here where the people you observe are not necessarily the same set as the people you play against, I think I just misunderstood what you wanted in your write-up. Do you want me to make the evidence neighborhood size a user-configurable parameter? And is there a reason not to allow it to go up to 24? It shouldn't be too hard to adjust the choice logic for different numbers of neighbors and risk attitude scale (and we'll need to do that anyway if they don't always match).

quadrismegistus reviewed (approved):
LGTM!

LaraBuchak commented:
Thanks! I was thinking it would be good to leave it at 8 because our risk-attitudes are defined discretely in terms of how many out of 8 neighbors you need to play hawk. But I suppose 24 is a multiple of 8, so it will translate cleanly (and 4 is at least a factor of 8, so there will still be 4 discrete types, even if r = 0 and r = 1 behave the same).

Let’s keep our risk attitudes ranging from 0 to 8 (or -1, when we put that in), but for the case of 24 “observed” neighbors, have the agent play HAWK when #neighbors-dove/3 > r, and for the case of 4 observed neighbors, have the agent play HAWK when #neighbors-dove x 2 > r. This way the r values will mean the same thing when varying the size of the observations neighborhood.

Does that sound good / make sense?

All best,
Lara

From: Rebecca Sutton Koeser ***@***.***>
Date: Wednesday, November 8, 2023 at 4:45 PM
To: Princeton-CDH/simulating-risk ***@***.***>
Cc: Lara M. Buchak ***@***.***>, Mention ***@***.***>
Subject: Re: [Princeton-CDH/simulating-risk] More options for hawk/dove neighborhoods (PR #46)

@rlskoeser commented on this pull request.

________________________________

In simulatingrisk/hawkdove/model.py<https://github.com/Princeton-CDH/simulating-risk/pull/46#discussion_r1387233658>:

>          hawk_odds=0.5,

     ):

         super().__init__()

         # assume a fully-populated square grid

         self.num_agents = grid_size * grid_size

-        # mesa get_neighbors supports moore neighborhood (include diagonals)

-        # and von neumann (exclude diagonals)

-        self.include_diagonals = include_diagonals

+        check_neighborhood_size(play_neighborhood)

+        self.play_neighborhood = play_neighborhood

+        # since we define risk attitudes from 0-8 or 0-4,

+        # the neighborhood for choosing what to play cannot be bigger than 8;

+        # use the smaller of the play neighborhood or 8

+        self.choose_neighborhood = min(8, play_neighborhood)

@LaraBuchak<https://github.com/LaraBuchak> thanks for responding. I wondered about calling it "observed neighborhood" (relates to seen / myopia) but wondered if that was ambiguous with the neighbors used for deciding how to adjust risk attitudes, which is a different kind of observation.

I have a sense of the potential interesting cases here where the people you observe are not necessarily the same set as the people you play against, I think I just misunderstood what you wanted in your write-up. Do you want me to make the evidence neighborhood size a user-configurable parameter? And is there a reason not to allow it to go up to 24? It shouldn't be too hard to adjust the choice logic for different numbers of neighbors and risk attitude scale (and we'll need to do that anyway if they don't always match).

—
Reply to this email directly, view it on GitHub<https://github.com/Princeton-CDH/simulating-risk/pull/46#discussion_r1387233658>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BBQKLGOL2GHLCEZP4VUWFH3YDP4PVAVCNFSM6AAAAAA7B4RHFSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOMRRGMZDSNRTGM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

rlskoeser commented on simulatingrisk/hawkdove/model.py:
```diff
@@ -66,18 +66,50 @@ def initial_choice(self, hawk_odds=None):
+        (uses `observed_neighborhood` size from model)"""
+        return len([n for n in self.observed_neighbors if n.last_choice == Play.DOVE])
+
+    @property
+    def proportional_num_dove_neighbors(self):
+        """adjust the number of dove neighbors based on ratio between
+        play neighborhood and observed neighborhood, to scale observations
+        to the range of agent risk level."""
+        ratio = (
+            min(self.model.max_risk_level, self.model.play_neighborhood)
+            / self.model.observed_neighborhood
+        )
+        # always round to an integer
+        return round(ratio * self.num_dove_neighbors)
```
@LaraBuchak here's the new logic for adjusting observed neighborhood so it matches the range of risk attitudes

rlskoeser commented:
@LaraBuchak that all makes sense! I've made observed neighborhood size configurable, adjusted the logic for choosing what to play, and updated the simrisk.pulcloud instance with this version of the code so you can review how the simulations look there.

rlskoeser commented:
@LaraBuchak thanks for catching that problem with the risk level sizing related to neighborhoods - I've updated so we always set risk levels 0-8 no matter the other parameters, and I think I've updated everything related to that, including charts and text in the readme. You should be able to see it now on  https://simrisk.pulcloud.io/hawkdove-multiple 

Please let me know if you see any other problems or errors.
